### ISSUE TITLE
Merge TuhiGui into this repository

### ISSUE THREAD

**Issue by whot** [tuhiproject/tuhi#140]
I've been working on a [tuhigui](http://github.com/whot/tuhigui), which is a GTK GUI for Tuhi.

I think this one should **not** stay a separate project but rather be merged into Tuhi itself, for a number of reasons:
- Tuhi is a DBus server based on the idea that "someone" will write GUIs for each desktop environment. In the last 18 months, no-one was motivated to write *any* GUI, so I very much doubt a proliferation of GUIs will start to exist.
- Tuhi cannot be flatpaked properly because it's not a GUI application. Were TuhiGui a separate app, we'd have to flatpak that one but include Tuhi in it to make sure the server is available.
- Name confusion: I don't want Tuhi to be the server and then have a separate name for the GUI that users will use. It should all just be Tuhi.
- The split between libratbag and piper is a disaster, with packages never matching git and the other way round. I don't want to recreate that here.
- I don't think there's a good argument for shipping Tuhi as its own package anyway. We don't even install `kete` right now. What's the point of a DBus server when there are no tools to actually interact with it out of the box.

**Comment by whot** [comment]
Addendum: long term this **may** mean that we can drop the DBus indirection and make Tuhi just a python library. Which means we don't have to think about things like the DBus daemon storing some data  nondeterministically just in case no GUI ever connects to it while it's running.

**Comment by bentiss** [comment]
Ack from my point of view. We wrote this project nearly two years ago, and nobody stood up to write the UI. So having one UI is better than none, and the easiest it will be for us the better.

**Comment by muesli** [comment]
As someone who's depending on tuhi and its DBus interface, I really hope you reconsider this decision. It may be a bit more cumbersome to develop a GUI for tuhi, but at the same time it enables projects like mine to re-use the same interface and establishes a clear divider between view and data models (or frontend and backend). In my humble opinion it would be a shame to throw away all that extra work you already put in to get this far.

Shameless plug for anyone who's interested: penpal, a syncing daemon depending on tuhi: https://github.com/muesli/penpal

**Comment by whot** [comment]
> Shameless plug for anyone who's interested: penpal, a syncing daemon depending on tuhi: 

Based on the README there - why not merge the functionality into TuhiGui? Especially the animation bit would be nice to have in the GUI. IIRC the Wacom Inkspace app provides sliders to export an SVG up to a certain point in the drawing, I'm sure something similar could be implemented for tuhigui as well.

**Comment by muesli** [comment]
For one, I don't actually want this to be a GUI application but a daemon that just keeps syncing new drawings automatically. This probably will keep running on a Raspberry Pi without an X server. Also, this is written in Go, so we can't just merge it over to a Python application.

Other than that: a slider like the Inkspace app provides should be fairly straight forward to integrate. Which reminds me I wanted to file another ticket, as it seems that `TOffset` is always missing in the JSON data. Is this something only the live mode provides?

**Comment by whot** [comment]
There are two parts to having a DBus daemon: there's the technical side of having the separation between the sync bits and the actual user of the bits - which allows for python tuhi and a go application in your case. The GObject integration we have *may* allow us to drop the DBus server and still get the same functionality we have right now if everything is in-tree. That's a purely technical question though and not as relevant.

The real killer is IMO the maintenance cost of having a DBus server separately from the primary user application (the GUI). Because this implies separate packages in distributions, error handling in the GUI when the server isn't available, documentation for how to start the server separately, and handling all the bug reports that come from users not starting the server correctly and thus the GUI failing, etc. This is a massive PITA for the libratbag project, doubly so because there the API evolves and you get random failures in the client because the Piper flatpak doesn't work with whatever package the distribution ships. I really don't want to have to deal with that again, especially for a project that will have fairly low priority.

The tuhigui flatpak works around this by starting Tuhi as it starts the GUI, thus guaranteeing that a) it's always there and b) users don't have to figure out how to start a separate server. That really is the big win here.

`penpal` is written in Go, but it's LOCC is what - less than 1000? And afaict it's not project that's been around for years either, the commit log is a day old. It's not unsurmountable to write the same thing in Python, should the DBus daemon disappear. The other option would be for `penpal` to use Tuhi as subproject and do the same thing that tuhigui does now - start the Tuhiserver as it starts. Means you can't run penpal and tuhigui at the same time but that's even more niche than anything we need to care about.

**Comment by muesli** [comment]
Just making sure I'm understanding what you're saying here: your plan is to drop the DBus server, but still have a separate process for `tuhiserver` and `tuhigui`? What form of IPC are you planning between `tuhiserver` and the frontend?

As I've said, I'm fully aware that this is a lot more cumbersome to maintain, even though I still think it's the right approach. The age or effort put into `penpal` isn't really the point here, either. While you're certainly right that it's rather unlikely that anyone would want to access their Bamboo device from multiple applications, the main benefit, as I see it, is having a central place to communicate with the device and a language agnostic way to access it.

**Comment by whot** [comment]
Just finished #145, so that may have some of the answers in code, irrefutably ;)

Right now, we still have `tuhi-gui` and `tuhi-server` and the new `tuhi` command starts both as separate processes, with DBus in between. 

I tried getting this into one process, but afaict the various `sync` calls we use in the dbus bindings will hang things, so... meh for now, it's good enough.

As for the case of dropping DBus - we're using GTK for the GUI and GObject in the tuhi server anyway. So pretty much all of the dbus bits could be replaced with GObject properties and signals and we'd get the same results as we have now, without a DBus intermediate layer. The data connection is already on a separate thread anyway.

Also, to clarify: I am not disagreeing with you that the split is the right approach, not at all (which is why we started as DBus server instead of a python module). My argument is whether we can *afford* that split, given that the repo has largely been languishing for a year and a bit and I have no idea how much time I'll have in the future to work on it. Having said that, right now removing the DBus server would be more work for little benefit, so it'll stay around for a bit longer just for that reason alone :)
